### ISSUE TITLE
Global errorChannel Listeners Issue in Exception Scenarios 

### ISSUE THREAD

**Issue by Mrc0113** [spring-cloud/spring-cloud-stream#2066]
We are seeing an issue where the Binder's Error MessageHandler that is handed to the framework is not being called if a developer throws an exception in their own custom code listening on the "errorChannel" via a `@ServiceActivator`. My assumption is that this issue is due to the MessageHandlers being executed in serial and the framework not catching an exception from the custom code and allowing the next MessageHandler to execute it's logic. Note my understanding is that the MessageHanders are subscribed to the errorChannel using the PubSub pattern so I would expect them all to execute and be decoupled from each other.  

This can be reproduced with the Solace binder (and I assume the Rabbit binder as they follow the same pattern) using the code below. You will see the `processError` code executes, throws an exception and then the Solace/Rabbit error handler is not invoked. If you then change the `processError` method to not throw an exception you will see that the Solace/Rabbit error handler is indeed invoked. 

```
@SpringBootApplication
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class);
	}

	@Bean
	public Function<String, String> processDropoffRideAverages() {
		return tsup -> {
			System.out.println("Received tsup:" + tsup);
			throw new RuntimeException("oh no!");
		};
	}

	@ServiceActivator(inputChannel = "errorChannel")
	private void processError(ErrorMessage errMsg) {
		System.out.println("processError: Received on global errorChannel:" + errMsg);
		throw new RuntimeException("Exception from global error handler!");
	}
}
```

*Note that best practice would be to code defensively and not have the custom error handling code throw an exception but error happen :)

**Comment by Mrc0113** [comment]
@olegz this is the issue that we discussed previously. Thanks for your continued guidance and support of the community!

**Comment by olegz** [comment]
Hey @Mrc0113 
I know you and I talked about it extensively, but I guess I am still missing something. 
Executing your code with rabbit and her e is what I get:
```
@ServiceActivator(inputChannel = "errorChannel")
private void processError(ErrorMessage errMsg) {
	System.out.println("processError: Received on global errorChannel:" + errMsg);
//		throw new RuntimeException("Exception from global error handler!");
}
```
stack trace
```
2021-01-23 08:47:55.893 ERROR 19574 --- [Wm2zouk3e--zA-1] o.s.integration.handler.LoggingHandler   : org.springframework.messaging.MessageHandlingException: error occurred in message handler [org.springframework.cloud.stream.function.FunctionConfiguration$FunctionToDestinationBinder$1@50d0e9fd]; nested exception is java.lang.RuntimeException: oh no!, failedMessage=GenericMessage [payload=byte[3], headers={amqp_receivedDeliveryMode=NON_PERSISTENT, amqp_receivedExchange=, amqp_deliveryTag=1, deliveryAttempt=3, amqp_consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, amqp_redelivered=false, target-protocol=amqp, amqp_receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, id=bddb0aae-f866-a5d8-0354-a60a49f8f54a, amqp_consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, sourceData=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), contentType=application/json, timestamp=1611388072877}]
	at org.springframework.integration.support.utils.IntegrationUtils.wrapInHandlingExceptionIfNecessary(IntegrationUtils.java:192)
	. . .
Caused by: java.lang.RuntimeException: oh no!
	at org.springone.function.core.FunctionApplication.lambda$0(FunctionApplication.java:46)
	at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.invokeFunctionAndEnrichResultIfNecessary(SimpleFunctionRegistry.java:733)
	at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.invokeFunction(SimpleFunctionRegistry.java:689)
	. . .
	... 27 more

2021-01-23 08:47:55.894  INFO 19574 --- [Wm2zouk3e--zA-1] o.s.i.h.s.MessagingMethodInvokerHelper   : Overriding default instance of MessageHandlerMethodFactory with provided one.
processError: Received on global errorChannel:ErrorMessage [payload=org.springframework.messaging.MessageHandlingException: error occurred in message handler [org.springframework.cloud.stream.function.FunctionConfiguration$FunctionToDestinationBinder$1@50d0e9fd]; nested exception is java.lang.RuntimeException: oh no!, failedMessage=GenericMessage [payload=byte[3], headers={amqp_receivedDeliveryMode=NON_PERSISTENT, amqp_receivedExchange=, amqp_deliveryTag=1, deliveryAttempt=3, amqp_consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, amqp_redelivered=false, target-protocol=amqp, amqp_receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, id=bddb0aae-f866-a5d8-0354-a60a49f8f54a, amqp_consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, sourceData=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), contentType=application/json, timestamp=1611388072877}], headers={amqp_raw_message=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), id=96c690d9-519c-eec4-c3e0-38d98130ef5b, sourceData=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), timestamp=1611388075891}] for original GenericMessage [payload=byte[3], headers={amqp_receivedDeliveryMode=NON_PERSISTENT, amqp_receivedExchange=, amqp_deliveryTag=1, deliveryAttempt=3, amqp_consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, amqp_redelivered=false, target-protocol=amqp, amqp_receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, id=bddb0aae-f866-a5d8-0354-a60a49f8f54a, amqp_consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, sourceData=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), contentType=application/json, timestamp=1611388072877}]
2021-01-23 08:47:55.898  INFO 19574 --- [Wm2zouk3e--zA-1] o.s.a.r.c.CachingConnectionFactory       : Attempting to connect to: [localhost:5672]
```
==
And then with:
```
@ServiceActivator(inputChannel = "errorChannel")
private void processError(ErrorMessage errMsg) {
	System.out.println("processError: Received on global errorChannel:" + errMsg);
	throw new RuntimeException("Exception from global error handler!");
}
```
stack trace
```
2021-01-23 08:47:55.893 ERROR 19574 --- [Wm2zouk3e--zA-1] o.s.integration.handler.LoggingHandler   : org.springframework.messaging.MessageHandlingException: error occurred in message handler [org.springframework.cloud.stream.function.FunctionConfiguration$FunctionToDestinationBinder$1@50d0e9fd]; nested exception is java.lang.RuntimeException: oh no!, failedMessage=GenericMessage [payload=byte[3], headers={amqp_receivedDeliveryMode=NON_PERSISTENT, amqp_receivedExchange=, amqp_deliveryTag=1, deliveryAttempt=3, amqp_consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, amqp_redelivered=false, target-protocol=amqp, amqp_receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, id=bddb0aae-f866-a5d8-0354-a60a49f8f54a, amqp_consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, sourceData=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), contentType=application/json, timestamp=1611388072877}]
	at org.springframework.integration.support.utils.IntegrationUtils.wrapInHandlingExceptionIfNecessary(IntegrationUtils.java:192)
	. . .
Caused by: java.lang.RuntimeException: oh no!
	at org.springone.function.core.FunctionApplication.lambda$0(FunctionApplication.java:46)
	at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.invokeFunctionAndEnrichResultIfNecessary(SimpleFunctionRegistry.java:733)
	at org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$FunctionInvocationWrapper.invokeFunction(SimpleFunctionRegistry.java:689)
	. . .
	... 27 more

2021-01-23 08:47:55.894  INFO 19574 --- [Wm2zouk3e--zA-1] o.s.i.h.s.MessagingMethodInvokerHelper   : Overriding default instance of MessageHandlerMethodFactory with provided one.
processError: Received on global errorChannel:ErrorMessage [payload=org.springframework.messaging.MessageHandlingException: error occurred in message handler [org.springframework.cloud.stream.function.FunctionConfiguration$FunctionToDestinationBinder$1@50d0e9fd]; nested exception is java.lang.RuntimeException: oh no!, failedMessage=GenericMessage [payload=byte[3], headers={amqp_receivedDeliveryMode=NON_PERSISTENT, amqp_receivedExchange=, amqp_deliveryTag=1, deliveryAttempt=3, amqp_consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, amqp_redelivered=false, target-protocol=amqp, amqp_receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, id=bddb0aae-f866-a5d8-0354-a60a49f8f54a, amqp_consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, sourceData=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), contentType=application/json, timestamp=1611388072877}], headers={amqp_raw_message=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), id=96c690d9-519c-eec4-c3e0-38d98130ef5b, sourceData=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), timestamp=1611388075891}] for original GenericMessage [payload=byte[3], headers={amqp_receivedDeliveryMode=NON_PERSISTENT, amqp_receivedExchange=, amqp_deliveryTag=1, deliveryAttempt=3, amqp_consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, amqp_redelivered=false, target-protocol=amqp, amqp_receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, id=bddb0aae-f866-a5d8-0354-a60a49f8f54a, amqp_consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, sourceData=(Body:'[B@24e1dc6b(byte[3])' MessageProperties [headers={}, contentLength=0, receivedDeliveryMode=NON_PERSISTENT, redelivered=false, receivedExchange=, receivedRoutingKey=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA, deliveryTag=1, consumerTag=amq.ctag-uU77k9JY2o0Ezp2deg_CGA, consumerQueue=processDropoffRideAverages-in-0.anonymous.dgzxNi6MSWm2zouk3e--zA]), contentType=application/json, timestamp=1611388072877}]
2021-01-23 08:52:04.563  WARN 20363 --- [mexT2mHfIC1FQ-1] s.a.r.l.ConditionalRejectingErrorHandler : Execution of Rabbit message listener failed.

org.springframework.amqp.rabbit.support.ListenerExecutionFailedException: Listener threw exception
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.wrapToListenerExecutionFailedExceptionIfNeeded(AbstractMessageListenerContainer.java:1705) ~[spring-rabbit-2.3.0-M2.jar:2.3.0-M2]
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:1595) ~[spring-rabbit-2.3.0-M2.jar:2.3.0-M2]
	...
Caused by: org.springframework.messaging.MessageHandlingException: error occurred during processing message in 'MethodInvokingMessageProcessor' [org.springframework.integration.handler.MethodInvokingMessageProcessor@242df928]; nested exception is java.lang.RuntimeException: Exception from global error handler!
	at org.springframework.integration.support.utils.IntegrationUtils.wrapInHandlingExceptionIfNecessary(IntegrationUtils.java:192) ~[spring-integration-core-5.4.0-M2.jar:5.4.0-M2]
	at org.springframework.integration.handler.MethodInvokingMessageProcessor.processMessage(MethodInvokingMessageProcessor.java:111) ~[spring-integration-core-5.4.0-M2.jar:5.4.0-M2]
	at org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:105) ~[spring-integration-core-5.4.0-M2.jar:5.4.0-M2]
	...
Caused by: java.lang.RuntimeException: Exception from global error handler!
	at org.springone.function.core.FunctionApplication.processError(FunctionApplication.java:53) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	...
	... 45 common frames omitted
```

What am I missing?

**Comment by olegz** [comment]
Digging a bit more I see that global error channel has two subscribers [LoggingHandler, ServiceActivatingHandler] - in that order. And that matches the stack trace where first the LoggingHandler is called and then the actual error handler as ServiceActivator subscribed to 'errorChannel'. 
Now, experimenting further, if I add another handler to the "errorChannel" and throw the exception in each, like so:
```
@ServiceActivator(inputChannel = "errorChannel")
private void processErrorA(ErrorMessage errMsg) {
	System.out.println("processError: Received on global errorChannel - 1:" + errMsg);
	throw new RuntimeException("Exception from global error handler!");
}
@ServiceActivator(inputChannel = "errorChannel")
private void processErrorB(ErrorMessage errMsg) {
	System.out.println("processError: Received on global errorChannel - 2:" + errMsg);
	throw new RuntimeException("Exception from global error handler!");
}
```

. . . then I do see the behavior that resembles what you are describing. Basically the subscribers are invoked sequentially and as soon as one throws an exception, no other subscribers are invoked. So basically with the exception I only see "SA Global: error - 1". Without i see "SA Global: error - 1" and "SA Global: error - 2".

Is that what you were saying? Perhaps that was the bit that was missing from your code?

**Comment by olegz** [comment]
One other possibility why I may not see it is that perhaps in Solace you do have some additional listener, at which point what you are describing makes perfect sense.
In any event, I was able to reproduce the behavior you are describing so need to figure out the way to fix it since most of it is actually happening on SI side.

**Comment by olegz** [comment]
@Mrc0113 so, after playing with it some more, I am now convinced that may be the reason why I don't see the behavior as you describe is because there may be some additional error handler in Solace that does not exist in Rabbit. That said, by simply adding another subscriber to the error channel reproduces the issue.
Easy fix, so will apply it shortly.
However, you can also do it now with a simple workaround (temporary).
Basically get access to the ApplicationContext and than:
```
PublishSubscribeChannel errorChannel = context.getBean("errorChannel", PublishSubscribeChannel.class);
errorChannel.setIgnoreFailures(true);
```

**Comment by Mrc0113** [comment]
> One other possibility why I may not see it is that perhaps in Solace you do have some additional listener, at which point what you are describing makes perfect sense.
> In any event, I was able to reproduce the behavior you are describing so need to figure out the way to fix it since most of it is actually happening on SI side.

@olegz Correct, I believe this is exactly the issue. The Solace Binder has an error handler that get handed to the framework [here](https://github.com/SolaceProducts/solace-spring-cloud/blob/69586d719162fafd151c67450bce871e904a0c2d/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder/src/main/java/com/solace/spring/cloud/stream/binder/SolaceMessageChannelBinder.java#L127) and I assume gets registered by the framework as a listener on the global error channel. However it must end up subscribing after the custom `@ServiceActivator` which is why it fails to be executed if the code in the case of an exception thrown on a previous subscriber. 

Thanks for looking into this and for the temporary workaround!
